openfecli/commands/gather.py

```diff
@@ -141,24 +142,77 @@ def dp2(v: float) -> str:
             DG2_mag, DG2_unc = vals['solvent']
             if not ((DG1_mag is None) or (DG2_mag is None)):
                 # DDG(2,1)bind = DG(1->2)complex - DG(1->2)solvent
-                DDGbind = dp2((DG1_mag - DG2_mag).m)
-                bind_unc = dp2(np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m]))))
+                DDGbind = (DG1_mag - DG2_mag).m
+                bind_unc = np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m])))
         if 'solvent' in vals and 'vacuum' in vals:
             DG1_mag, DG1_unc = vals['solvent']
             DG2_mag, DG2_unc = vals['vacuum']
             if not ((DG1_mag is None) or (DG2_mag is None)):
-                DDGhyd = dp2((DG1_mag - DG2_mag).m)
-                hyd_unc = dp2(np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m]))))
+                DDGhyd = (DG1_mag - DG2_mag).m
+                hyd_unc = np.sqrt(np.sum(np.square([DG1_unc.m, DG2_unc.m])))
+
+        DDGs.append((*ligpair, DDGbind, bind_unc, DDGhyd, hyd_unc))
+
+    MLEs = []
+    # 4b) perform MLE
+    g = nx.DiGraph()
+    nm_to_idx = {}
+    DDGbind_count = 0
+    for ligA, ligB, DDGbind, bind_unc, DDGhyd, hyd_unc in DDGs:
+        if DDGbind is None:
+            continue
+        DDGbind_count += 1
+
+        # tl;dr this is super paranoid, but safer for now:
+        # cinnabar seems to rely on the ordering of values within the graph
+        # to correspond to the matrix that comes out from mle()
+        # internally they also convert the ligand names to ints, which I think
+        # has a side effect of giving the graph nodes a predictable order.
+        # fwiw this code didn't affect ordering locally
+        try:
+            idA = nm_to_idx[ligA]
+        except KeyError:
+            idA = len(nm_to_idx)
+            nm_to_idx[ligA] = idA
+        try:
+            idB = nm_to_idx[ligB]
+        except KeyError:
+            idB = len(nm_to_idx)
+            nm_to_idx[ligB] = idB
+
+        g.add_edge(
+            idA, idB, calc_DDG=DDGbind, calc_dDDG=bind_unc,
+        )
+    if DDGbind_count > 2:
+        idx_to_nm = {v: k for k, v in nm_to_idx.items()}
+
+        f_i, df_i = mle(g, factor='calc_DDG')
+        df_i = np.diagonal(df_i) ** 0.5
 
-        name = ", ".join(ligpair[::-1])
+        for node, f, df in zip(g.nodes, f_i, df_i):
+            ligname = idx_to_nm[node]
+            MLEs.append((ligname, f, df))
+
+    output.write('measurement\ttype\tligand_i\tligand_j\testimate (kcal/mol)'
+                 '\tuncertainty (kcal/mol)\n')
+    # 5a) write out MLE values
+    for ligA, DG, unc_DG in MLEs:
+        DG, unc_DG = dp2(DG), dp2(unc_DG)
+        output.write(f'DGbind({ligA})\tDG(MLE)\tZero\t{ligA}\t{DG}\t{unc_DG}\n')
```

Comment:
@hannahbaumann this will print a line like:

```
measurement type ligand_i ligand_j estimate (kcal/mol) uncertainty (kcal/mol)
DGbind(lig_1234) DG(mle) Zero lig_1234 -4.2 0.23
```

would this make sense to you?  "zero" is meant to be the zero state reference point, not sure if there's a better label for that

Reply:
It's not a zero state dG though? Unless you normalise against experiment your reference state is the first ligand passed to your MLE

Reply:
In any case, if it's possible to inspect the original SMCs for stored experimental values this becomes a lot more useful. Otherwise what I would do is return a list of values with the leading entry being the 0 kcal/mol entry. That way it's clear what the central point is.

Reply:
Yes, I think from that print statement I wouldn't know what zero means, so what the reference is and therefore what meaning the ("relative") dG has. Or is that described elsewhere?